entity-framework/core/performance/efficient-querying.md

```diff
@@ -183,6 +183,43 @@ Here are the results for a benchmark comparing tracking vs. no-tracking behavior
 
 Finally, it is possible to perform updates without the overhead of change tracking, by utilizing a no-tracking query and then attaching the returned instance to the context, specifying which changes are to be made. This transfers the burden of change tracking from EF to the user, and should only be attempted if the change tracking overhead has been shown to be unacceptable via profiling or benchmarking.
 
+## Efficient pagination
+
+Pagination refers to retrieving results in pages, rather than all at once; this is typically done for large resultsets, where a user interface is shown that allows the user to navigate to the next or previous page of the results.
+
+A common way to implement pagination with databases is to use the `Skip` and `Take` (`OFFSET` and `LIMIT` in SQL). Given a a page size of 10 results, the third page can be fetched with EF Core as follows:
+
+[!code-csharp[Main](../../../samples/core/Performance/Program.cs#OffsetPagination)]
+
+Unfortunately, while this technique is very intuitive, it also has some severe shortcomings:
+
+1. The database must still process the first 20 entries, even if they aren't returned to the application; this creates possibly significant computation load that increases with the number of rows being skipped.
+2. If any updates occur concurrently, your pagination may end up skipping certain entries or showing them twice. For example, if an entry is removed as the user is moving from page 2 to 3, the whole resultset "shifts up", and one entry would be skipped.
+
+The recommended alternative to offset-based pagination - sometimes called *keyset pagination* - is simply to remember the details of the last entry fetched (e.g. its ID, assuming you're ordering by that), and to ask for the next rows after that row. For example, assuming the last entry in the last page we fetched had an ID value of 55, we'd simply do the following:
+
+[!code-csharp[Main](../../../samples/core/Performance/Program.cs#KeySetPagination)]
+
+Assuming an index is defined on `PostId`, this query is very efficient, and also isn't sensitive to any concurrent changes happening in lower Id values.
+
+Keyset pagination is appropriate for pagination interfaces where the user navigates forwards and backwards, but does not support random access, where the user can jump to any specific page. Random access pagination typically requires using offset pagination as explained above.
```

Comment:
Is it worth suggesting a combination of both? For example, use an offset-based pagination for forward one or maybe a few pages, with skip/take to reset to a randomly jumped to page?

Reply:
Good idea, will do.